### PR TITLE
Add OnSecurityStampReplacePrincipal event

### PR DESCRIPTION
**Pull request by HaoK**
Fixes https://github.com/aspnet/Identity/issues/958

Adds a new event called from the default security stamp validator which allows tweaking the user principal regeneration using the old principal.

Not super happy with the name (or the fact that the event lives at the top level of IdentityOptions)

cc @divega @blowdart

### CODE REVIEW

**Comment by brockallen** [comment]
> or the fact that the event lives at the top level of IdentityOptions

This is why there should be a security stamp options (maybe in 2.0.0 when you can add a breaking change).

**Comment by HaoK** [comment]
@Tratcher mind reviewing as well since you are the only other one with some context :)

**Comment by brockallen** [comment]
After a quick glance, it looks ok to me (FWIW)

**Comment by Tratcher** [comment]
What's the other event where the user can copy claims from the external cookie to the identity cookie?

**Comment by HaoK** [comment]
CreatingTicket I believe is the other OAuth event that can be hooked

**Comment by Tratcher** [comment]
? Wasn't there an event at the identity layer?

**Comment by HaoK** [comment]
Nope, shuttling claims around was always done directly via cookie events.  We added a method which took care of some of the token refreshing, but identity doesn't expose any cookie related events, it just exposes the cookie options directly (until this PR anyways...)

File: src/Microsoft.AspNetCore.Identity/SecurityStampReplacingPrincipalContext.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+{
+    /// <summary>
+    /// Used to pass information during the SecurityStamp validation event.
+    /// </summary>
+    public class SecurityStampReplacingPrincipalContext
+    {
+        /// <summary>
+        /// The principal contained in the current cookie.
+        /// </summary>
+        public ClaimsPrincipal Current { get; set; }
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Identity/SecurityStampReplacingPrincipalContext.cs]: CurrentPrincipal and NewPrincipal?

File: src/Microsoft.AspNetCore.Identity/IdentityOptions.cs [hunk 2]
```diff
@@ -74,5 +75,10 @@ public class IdentityOptions
         /// The <see cref="TimeSpan"/> after which security stamps are re-validated.
         /// </value>
         public TimeSpan SecurityStampValidationInterval { get; set; } = TimeSpan.FromMinutes(30);
+
+        /// <summary>
+        /// Invoked when the default security stamp validator replaces the user's ClaimsPrincipal in the cookie.
+        /// </summary>
+        public Func<SecurityStampReplacingPrincipalContext, Task> OnSecurityStampReplacingPrincipal { get; set; }
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Identity/IdentityOptions.cs]: Refreshing?

**Comment by HaoK** [comment]
Sure, updated to suggested names, New/Current Principal + Replacing => Refreshing

**Comment by HaoK** [comment]
aba7cf96b9ac7c910ee7db5e34fa7cf78516116a

**Pull request closed without merging**